R/JoinKeys.R

```diff
@@ -164,12 +164,92 @@ JoinKeys <- R6::R6Class( # nolint
       } else {
         cat("An empty JoinKeys object.")
       }
+      invisible(self)
+    },
+    #' @description
+    #' Sets the parents of the datasets.
+    #'
+    #' @param named_list Named (`list`) of the parents datasets.
+    #'
+    #' @return (`self`) invisibly for chaining
+    set_parents = function(named_list) {
+      for (dataset in names(named_list)) {
+        checkmate::assert(
+          checkmate::check_null(self$get_parent(dataset)),
+          checkmate::check_true(
+            length(self$get_parent(dataset)) == 0 &&
+              length(named_list[[dataset]]) == 0
+          ),
+          checkmate::check_true(self$get_parent(dataset) == named_list[[dataset]]),
+          "Please check the difference between provided datasets parents and provided join_keys parents."
+        )
+        if (is.null(self$get_parent(dataset))) {
+          private$parents[[dataset]] <- named_list[[dataset]]
+        }
+      }
+      invisible(self)
+    },
+    #' @description
+    #' Gets the parent of the desired dataset.
+    #'
+    #' @param dataname (`character`) name of the dataset.
+    #' @return (`character`) the parent of the desired dataset
+    get_parent = function(dataname) {
+      private$parents[[dataname]]
+    },
+    #' @description
+    #' Gets the parents of the datasets.
+    #'
+    #' @return (`list`) A named list of the parents of all datasets
+    get_parents = function() {
+      private$parents
+    },
+    #' @description
+    #' Updates the keys of the datasets based on the parents.
+    #'
+    #' @return (`self`) invisibly for chaining
+    update_keys_given_parents = function() {
+      datanames <- names(self$get())
+      duplicate_pairs <- list()
+      for (d1 in datanames) {
+        d1_pk <- self$get(d1, d1)
+        d1_parent <- self$get_parent(d1)
+        for (d2 in datanames) {
+          if (paste(d2, d1) %in% duplicate_pairs) {
+            next
+          }
+          if (length(self$get(d1, d2)) == 0) {
+            d2_parent <- self$get_parent(d2)
+            d2_pk <- self$get(d2, d2)
+
+            fk <- if (identical(d1, d2_parent)) {
+              # first is parent of second -> parent keys -> first keys
+              d1_pk
+            } else if (identical(d1_parent, d2)) {
+              # second is parent of first -> parent keys -> second keys
+              d2_pk
+            } else if (identical(d1_parent, d2_parent) && length(d1_parent) > 0) {
+              # both has the same parent -> parent keys
+              self$get(d1_parent, d1_parent)
+            } else {
+              # cant find connection - leave empty
+              next
+            }
+            self$mutate(d1, d2, fk)
+            duplicate_pairs <- append(duplicate_pairs, paste(d1, d2))
+          }
+        }
```

Comment:
Since `mutate` changes keys to `dataset1` and `dataset2` in the same time, can we just make a simpler loop? If keys are taken from parents we don't need this `if` in the line 225, isn't it?

```suggestion
      datanames <- names(self$get())
      duplicate_pairs <- list()
      for (d1 in datanames) {
        d1_pk <- self$get(d1, d1)
        d1_parent <- self$get_parent(d1)
        if (length(d1_parent) > 0) {
          parent_pk <- self$get(d1_parent, d1_parent)
          self$mutate(d1, d1_parent, parent_pk)
        }
      }
      # check parent child relation
      private$check_parent_child()

      invisible(self)
```

Reply:
No, we need the double loop.
Imagine we have a `JoinKeys` object with 3 datasets inside, `ADSL`, `ADAE` and `ADTTE`. We define join_keys between `ADSL`  (the parent) and each of `ADAE` and `ADTTE` which turn out to be the same keys. In this case, the double loop creates join_keys between `ADAE` and `ADTTE` which are the same keys with the parent dataset. With the single loop, this won't be the case.